src/Storage/Sha1FileHashingService.php

```diff
@@ -0,0 +1,193 @@
+<?php
+
+namespace SilverStripe\Assets\Storage;
+
+use InvalidArgumentException;
+use League\Flysystem\FileNotFoundException;
+use League\Flysystem\Filesystem;
+use League\Flysystem\Util;
+use Psr\SimpleCache\CacheInterface;
+use SilverStripe\Assets\Flysystem\FlysystemAssetStore;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Core\Injector\Injector;
+
+/**
+ * Utility for computing and comparing unique file hash. All `$fs` parameters can either be:
+ * * an `AssetStore` constant VISIBILITY constant or
+ * * an actual `Filesystem` object.
+ *
+ * @internal This interface is not part of the official SilverStripe API and may be altered in minor releases.
+ */
+class Sha1FileHashingService implements FileHashingService
+{
+    use Configurable;
+    use Injectable;
+
+    /**
+     * Whetever Sha1FileHashingService should cache hash values by default.
+     * @var bool
+     * @config
+     */
+    private static $default_cachable = false;
```

Comment:
Caching will be disabled by default. Specific task like the MigrateFileTask can enabled caching on the fly as needed.

I could imagine this being enabled by default, but we don't have time to wrestle with this potential issue at this time, so it might be a SilverStripe 4.5 thing.

Reply:
I generally agree with keeping complexity low by using caching judiciously - it's a minefield of edge cases. We need to speed up file migration first and foremost. 

But: I've realised that by your 4.4 refactoring into filesystem helpers, even `File->exists()` check *always* compute hashes now (through `applyFileOnFilesystem()`). They didn't do this on 4.3: https://github.com/silverstripe/silverstripe-assets/blob/1.3.3/src/Flysystem/FlysystemAssetStore.php#L701. That'll slow down every file-based operation, incl. listing files in AssetAdmin.

Can we fix this without side effects? I don't see why an `exists()` check needs to check the file hash. Maybe we need "read" and "write" variations of `applyFileOnFilesystem`?

Reply:
Yeah ... that kind of suck. There's some methods on FileIDHelper that accept a `$strict` flag, that allows to ask for an explicit hash check. I ended up having to leave it on most of the time, because I kept knocking my head on edge cases. It's kind of annoying that you have to be aware of the underlying FileIDResolutionStrategy logic all the time.

A possible, alternative could be to do an alternative implementation of FileIDResolutionStrategy that assumes that one and only one FileResolution scheme is in use for each store. That could only be enabled after users run the file migration task. That FileResolutionStartegy wouldn't need to be so picky about hashes, so it would be a lot faster.

Reply:
As a benchmark, I looked at the time it takes asset-admin to list a folder with 20 3-MB-images.

When waiting for the page to be usable, it takes:
* 2.7 sec with hash caching enabled all the time
* 10.7 sec with hash caching disabled

Reply:
Raise this issue to look at performance concerns. https://github.com/silverstripe/silverstripe-assets/issues/279